Zend/zend_hash.c

```diff
@@ -459,12 +459,25 @@ ZEND_API void ZEND_FASTCALL _zend_hash_iterators_update(HashTable *ht, HashPosit
 	}
 }
 
-static zend_always_inline Bucket *zend_hash_find_bucket(const HashTable *ht, zend_string *key)
+/* To protect against HashDos attacks, we count collisions during the "find
+ * existing bucket with this key" phase of insertion operations. The check
+ * against MAX_COLLISIONS is only performed if *no* matching bucket is found,
+ * as that corresponds to insert operations (as opposed to updates).
+ *
+ * An important caveat of the implementation is that collisions will *not*
+ * be counted for add_new operations. find+add_new operations should be replaced
+ * with add_or_return operations.
+ */
+#define HT_MAX_COLLISIONS 1000
```

Comment:
What happens if a developer runs into this issue with an application without a malicious intention? Should this maybe a ini setting?

Or is this case really absolutely hypothetical (I don't have a in-depth understanding of the possibility of this happening outside actual DOS attempts) ?

Thank you!